plugins/user.py

```diff
@@ -0,0 +1,454 @@
+from will.plugin import WillPlugin
+from will.decorators import respond_to, require_settings
+from will import settings
+import pyotp
+import qrcode
+import base64
+import boto
+import boto.s3.connection
+from boto.s3.key import Key
+import cStringIO
+import datetime
+import urlparse
+from simplecrypt import encrypt, decrypt
+import pytz
+
+import logging
+log = logging.getLogger(__name__)
+
+
+# TODO: QR code over hipchat leaves the account vulnerable if the client renders the image and leaves it rendered in the chat history
+# TODO: Maybe vulnerable to brute force attacks, is there any rate limiting from client -> hipchat? or hipchat -> bot?
```

Comment:
Hipchat has ratelimits of 500 requests per 5 minutes.  The secret code changes every 30 seconds.